menas/src/main/scala/za/co/absa/enceladus/menas/services/DatasetService.scala

```diff
@@ -102,4 +102,150 @@ class DatasetService @Autowired() (datasetMongoRepository: DatasetMongoRepositor
     }
   }
 
+  override def importItem(item: Dataset, username: String): Future[Option[Dataset]] = {
+    getLatestVersionValue(item.name).flatMap {
+      case Some(version) => update(username, item.copy(version = version))
+      case None => super.create(item.copy(version = 1), username)
+    }
+  }
+
+  override def validateSingleImport(item: Dataset): Future[Validation] = {
+    val confRulesWithConnectedEntities = item.conformance.filter(_.hasEntityConnected)
+    val maybeSchema = datasetMongoRepository.getConnectedSchema(item.schemaName, item.schemaVersion)
+
+    val validations = super.validateSingleImport(item)
+    val validationsWithSchema = validateSchema(item.schemaName, item.schemaVersion, validations, maybeSchema)
+    val validationWithConnected = validateConnectedEntities(confRulesWithConnectedEntities, validationsWithSchema)
+    validateConformanceRules(item.conformance, maybeSchema, validationWithConnected)
+  }
+
+  private def validateConnectedEntities(confRulesWithConnectedEntities: List[ConformanceRule],
+                                    validationWithSchema: Future[Validation]): Future[Validation] = {
+    confRulesWithConnectedEntities.foldLeft(validationWithSchema) { (acc, cr) =>
+      val maybeConnectedEntity = cr.maybeConnectedEntityType.get match {
+        case "mappingTable" => datasetMongoRepository.getConnectedMappingTable(
+          cr.maybeConnectedEntityName.get,
+          cr.maybeConnectedEntityVersion.get
+        )
+      }
+
+      val issueMsg =
+        s"""Connected ${cr.maybeConnectedEntityType.get} ${cr.maybeConnectedEntityName.get}
+           | v${cr.maybeConnectedEntityVersion.get} could not be found""".stripMargin.replaceAll("[\\r\\n]", "")
+
+      for {
+        ce <- maybeConnectedEntity
+        accValidations <- acc
+      } yield accValidations.withErrorIf(ce.isEmpty, cr.maybeConnectedEntityType.get, issueMsg)
+    }
+  }
+
+  private def validateConformanceRules(conformanceRules: List[ConformanceRule],
+                                       maybeSchema: Future[Option[Schema]],
+                                       validations: Future[Validation]): Future[Validation] = {
+    val maybeFields = maybeSchema.map { case Some(x) => x.fields.flatMap(f => f.getAllChildren :+ f.getAbsolutePath()).toSet }
+    val accumulator = Tuple2(validations, maybeFields)
+
+    conformanceRules.foldLeft(accumulator) { case ((validation, currentColumns), conformanceRule) =>
+      conformanceRule match {
+        case cr: CastingConformanceRule =>
+          validateInAndOut(validation, currentColumns, cr)
+        case cr: NegationConformanceRule =>
+          validateInAndOut(validation, currentColumns, cr)
```

Comment:
Should these methods be directly on the Conformance Rules?

Reply:
I have been thinking about (trying a bit), to move it to conformance rules. But since it is built on top of features it does not make sense to pull it there. But to move it to a different file in the Menas maybe Menas/Services might do the trick of improving readability.

Maybe I could define on the ConformanceRules which type of validation they use, the same as "connected entities"?

Reply:
Let's do it in the future. Good enough for now.